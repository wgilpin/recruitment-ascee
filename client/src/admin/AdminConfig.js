import React from 'react';
import FetchData from '../common/FetchData';
import Styles from '../common/Styles';
import TableStyles from '../evidence/TableStyles';
import Alt from '../common/Alt'
import Confirm from '../common/Confirm';


/*
api required
/api/admin/get_congig/<str:key>
*/

const styles = {
  ...TableStyles.styles,
  ...Styles.styles,
  outer: {
    maxWidth: '400px',
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  textarea: {
    width: '300px',
    height: '300px',
    backgroundColor: '#111',
    color: 'white',
    margin: '8px',
    padding: '8px',
  },
  hr: {
    ...TableStyles.hr,
    color: 'grey',
    width: '350px',
  },
  input: {
    padding: '8px',
    margin: '8px',
    width: '300px',
    backgroundColor: '#111',
    color: 'white',
    height: '24px',
    borderWidth: '1px',
    borderColor: 'grey',
  },
  alt:{
    width: 'fit-content',
    marginLeft: 'auto',
    marginRight: 'auto',
  }
}
const templateKinds = {
  invite: 'Invitation Template',
  reject: 'Rejection Template',
}

export default class AdminConfig extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      mailCharacterId: null,
      mailCharacterName: '',
      mailTemplate: '',
      mailSubject: '',
      showConfirm: false,
      dirtyMailTemplate: false,
    };
    this.mailTemplate = React.createRef();
    this.mailSubject = React.createRef();
    this.mailKind = React.createRef();
  }

  componentDidMount() {
    new FetchData({ scope: 'mail/get_character' })
      .get()
      .then(({ info: { id, name }}) =>{
        this.setState({ mailCharacterId: id, mailCharacterName: name })
      })
    this.handleSelectTemplate();
  }

  handleSelectTemplate = () => {
    const kind = this.mailKind.current.value;
    new FetchData({ scope: `mail/template/${kind}` })
      .get()
      .then(({ info: { subject, text }}) =>{
        this.setState({
          mailKind: kind,
          mailSubject: subject,
          mailTemplate: text,
          dirtyMailTemplate: false,
        })
      })
  }

  doConfirmed = () => {
    window.location ='/api/mail/set_character';
  }

  handleSaveTemplate = () => {
    const template = {
      name: this.mailKind.current.value,
      subject: this.mailSubject.current.value,
      template: this.mailTemplate.current.value,
    }
    new FetchData({ scope: 'mail/template' })
      .put(template)
      .then (() => this.setState({ dirtyMailTemplate: false }))
  }

  handleEditTemplate = () => {
    this.setState({
      mailTemplate: this.mailTemplate.current.value,
      mailSubject: this.mailSubject.current.value,
      dirtyMailTemplate:
        this.mailTemplate.current.value.length > 0 &&
        this.mailSubject.current.value.length > 0,
    });
  }

  handleEditMail = () => this.setState({ showConfirm: true });

  render() {
    return [
      <div style={styles.outer}>
        <h2 style={styles.h2}>Mail User</h2>
        {this.state.mailCharacterId &&
          <Alt
            style={styles.alt}
            id={this.state.mailCharacterId}
            name={this.state.mailCharacterName}
          />
        }
        <button style={{ ...styles.secondaryButton, color: 'white'}} onClick={this.handleEditMail}>
          Edit
        </button>
        <hr style={styles.hr} />
        <h2 style={styles.h2}>Mail Template</h2>
        <select
          placeholder="Purpose"
          ref={this.mailKind}
          style={styles.input}
          onChange={this.handleSelectTemplate}
        >
          {Object.entries(templateKinds).map(([key, val]) => (
            <option
              value={key}
              selected={this.state.mailKind === key}>
                {val}
              </option>
          ))}
        </select>
        <div>
          <input
            style={styles.input}
            placeholder="  Subject Line"
            value={this.state.mailSubject}
            ref={this.mailSubject}
            onChange={this.handleEditTemplate}
          />
        </div>
        <textarea
          style={styles.textarea}
          value={this.state.mailTemplate}
          ref={this.mailTemplate}
          onChange={this.handleEditTemplate} />
        {this.state.dirtyMailTemplate &&
          <button style={styles.primaryButton} onClick={this.handleSaveTemplate}>
            Save
          </button>
        }
      </div>,
      this.state.showConfirm &&
        <Confirm
          text={`You will now be asked to log in the character you want to be the mail sender.
                This character needs to already have admin permission.` }
          onConfirm={this.doConfirmed}
          onClose={() => this.setState({ showConfirm: false })}
        />
    ];
  }
}
