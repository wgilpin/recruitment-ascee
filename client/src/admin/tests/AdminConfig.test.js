import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import AdminConfig from '../AdminConfig';

let mockFetch;

beforeEach(() => {
  const serverState = {
    get_character: {
      id: 1234,
      name: 'user 1',
    },
    template: {
      name: 'invite',
      subject: 'Test Rejection Email',
      text: "Ignore this - you haven't really been rejected",
    },
  };

  mockFetch = jest.fn().mockImplementation((scope, params) => {
    if (scope.scope === 'mail/get_character') {
      return Promise.resolve({
        info: serverState.get_character,
      });
    } else if (scope.scope.indexOf('mail/template') !== -1) {
      return Promise.resolve({
        info: serverState.template,
      });
    } else {
      throw new Error('Scope Not Valid');
    }
  });
});

describe('<AdminConfig>', () => {
  it('renders without crashing', () => {
    mount(<AdminConfig fetcher={mockFetch}/>);
  });

  it('matches snapshot', async () => {
    const adminConfig = mount(<AdminConfig fetcher={mockFetch} />);
    const tree = renderer.create(adminConfig);
    await Promise.resolve();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
