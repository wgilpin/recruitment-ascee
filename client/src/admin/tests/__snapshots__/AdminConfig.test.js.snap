// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminConfig> matches snapshot 1`] = `
<div
  style={
    Object {
      "marginLeft": "auto",
      "marginRight": "auto",
      "maxWidth": "400px",
    }
  }
>
  <h2>
    Mail User
  </h2>
  <div
    onClick={[Function]}
    style={
      Object {
        "cursor": null,
        "display": "table",
        "marginLeft": "auto",
        "marginRight": "auto",
        "padding": 8,
        "width": "fit-content",
      }
    }
  >
    <div
      style={
        Object {
          "display": "table-row",
        }
      }
    >
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
          }
        }
      >
        <img
          alt="Avatar"
          onClick={[Function]}
          src="https://image.eveonline.com/Character/1234_64.jpg"
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": "50%",
              "boxShadow": "1px 3px 1px #111",
              "height": 32,
              "width": 32,
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "display": "table-cell",
            "height": "24px",
            "marginBottom": "auto",
            "marginTop": "auto",
            "paddingLeft": "8px",
            "size": 14,
            "textAlign": "left",
            "verticalAlign": "middle",
          }
        }
      >
        user 1
      </div>
    </div>
  </div>
  <button
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#333",
        "borderColor": "#01799A",
        "borderRadius": "3px",
        "borderStyle": "none",
        "color": "white",
        "fontSize": "large",
        "margin": "12px",
        "padding": "6px",
        "width": "200px",
      }
    }
  >
    Edit
  </button>
  <hr
    style={
      Object {
        "color": "grey",
        "width": "350px",
      }
    }
  />
  <h2>
    Mail Template
  </h2>
  <select
    defaultValue=""
    onChange={[Function]}
    placeholder="Purpose"
    style={
      Object {
        "backgroundColor": "#111",
        "borderColor": "grey",
        "borderWidth": "1px",
        "color": "white",
        "height": "unset",
        "margin": "8px",
        "padding": "8px",
        "width": "300px",
      }
    }
  >
    <option
      value="invite"
    >
      Invitation Template
    </option>
    <option
      value="reject"
    >
      Rejection Template
    </option>
  </select>
  <div>
    <input
      onChange={[Function]}
      placeholder="  Subject Line"
      style={
        Object {
          "backgroundColor": "#111",
          "borderColor": "grey",
          "borderWidth": "1px",
          "color": "white",
          "height": "24px",
          "margin": "8px",
          "padding": "8px",
          "width": "300px",
        }
      }
      value="Test Rejection Email"
    />
  </div>
  <textarea
    onChange={[Function]}
    style={
      Object {
        "backgroundColor": "#111",
        "color": "white",
        "height": "300px",
        "margin": "8px",
        "padding": "8px",
        "width": "300px",
      }
    }
    value="Ignore this - you haven't really been rejected"
  />
</div>
`;
